compiler/transforms/fuse_streaming_regions.py

```diff
@@ -0,0 +1,168 @@
+        for o in op.body.block.ops:
+            # make copies of generic ops from producer region
+            if isinstance(o, stream.GenericOp):
+                rewriter.insert_op(
+                    producer_generic := stream.GenericOp(
+                        inputs=tuple(
+                            streaming_region_op.body.block.args[i.index]
+                            if isinstance(i, BlockArgument)
+                            and isinstance(i.type, stream.StreamType)
+                            else i
+                            for i in o.inputs
+                        ),
+                        body=rewriter.move_region_contents_to_new_regions(o.body),
+                        doc=o.doc,
+                        library_call=o.library_call,
+                        result_types=o.result_types,
+                    ),
+                    InsertPoint.at_end(streaming_region_op.body.block),
+                )
+            # do not use yield op from producer region
+            elif isinstance(o, stream.YieldOp):
+                continue
+            else:
+                raise RuntimeError("Can only fuse with generic bodies")
+        assert producer_generic is not None
+
+        consumer_generic = None
+        for o in user_op.body.block.ops:
+            # make copies of generic ops from consumer region
+            if isinstance(o, stream.GenericOp):
+                rewriter.insert_op(
+                    consumer_generic := stream.GenericOp(
+                        inputs=tuple(
+                            producer_generic.results[0]
+                            if index == obliterating_index
+                            else streaming_region_op.body.block.args[
+                                index + len(op.inputs) - 1
+                            ]
+                            for index in range(len(o.inputs))
+                        ),
+                        body=rewriter.move_region_contents_to_new_regions(o.body),
+                        doc=o.doc,
+                        library_call=o.library_call,
+                        result_types=o.result_types,
+                    ),
+                    InsertPoint.at_end(streaming_region_op.body.block),
+                )
+            elif isinstance(o, stream.YieldOp):
+                assert consumer_generic
+                rewriter.insert_op(
+                    stream.YieldOp(consumer_generic.results[0]),
+                    InsertPoint.at_end(streaming_region_op.body.block),
+                )
+
+            else:
+                raise RuntimeError("Can only fuse with generic bodies")
+
+        rewriter.replace_op(user_op, streaming_region_op)
+        rewriter.erase_matched_op()
+
+
+@dataclass(frozen=True)
+class FuseStreamingRegions(ModulePass):
+    name = "fuse-streaming-regions"
+
+    def apply(self, ctx: MLContext, op: builtin.ModuleOp) -> None:
+        PatternRewriteWalker(
+            FuseElementwisePattern(), apply_recursively=False
```

Comment:
Why can we not apply this pass recursively, now this will maximally fuse two operations, right? I think we should be able to fuse multiple operations together if we like, no?

Reply:
yes the apply recursively is now gone